### ISSUE TITLE
[BUG]  AdminNetworkPolicyIngressRule.Ports has two nil values?

### ISSUE THREAD

**Issue by fasaxc** [kubernetes-sigs/network-policy-api#247]
While reviewing our WIP implementation of ANP, I noticed that the Ports field is defined as 
```
	// Ports allows for matching traffic based on port and protocols.
	// This field is a list of ports which should be matched on
	// the pods selected for this policy i.e the subject of the policy.
	// So it matches on the destination port for the ingress traffic.
	// If Ports is not set then the rule does not filter traffic via port.
	//
	// Support: Core
	//
	// +optional
	// +kubebuilder:validation:MaxItems=100
	Ports *[]AdminNetworkPolicyPort `json:"ports,omitempty"`
```
which means that it has two/three "empty" values: 

* `(*[]AdminNetworkPolicyPort)(nil)`
* `&([]AdminNetworkPolicyPort(nil))`
* `&([]AdminNetworkPolicyPort{})`

It's not clear what they're each supposed to mean, "If Ports is not set then the rule does not filter traffic via port".  Clearly a nil pointer is "not set" but what about a pointer to an empty/nil slice?  Would those mean "match no ports" (which in turn is a bit ambiguous since not all protocols have ports)?

I suggest dropping the extra pointer indirection and interpreting any zero-length slice (nil or otherwise) as "do not filter on ports".  If we can't do that, we should at least document the meanings (where I think we should probably make them all mean "do not filter on ports").

**Comment by tssurya** [comment]
cc @astoycos

**Comment by danwinship** [comment]
For future reference, the rationale for the existing semantics is:

- `nil`/not-present _must_ mean "this has no effect on matching". Eg, if we add `Protocols` in the future, then the behavior when you don't specify `Protocols` must be the same as what the behavior was in the version of ANP where that field didn't even exist. eg, "don't restrict based on ICMP". So by extension, not specifying `Ports` should mean "don't restrict based on port".
- For array-valued fields, we had decided that 0-length should not be treated as a special case with non-standard behavior; it should just be the obvious extension of the non-0-length behavior. An ANP whose `Ports` is a 3-element array matches 3 ports. An ANP whose `Ports` is a 2-element array matches 2 ports. An ANP whose `Ports` is a 1-element array matches 1 port. Therefore an ANP whose `Ports` is a 0-element array should match 0 ports. Note that this doesn't imply the existence of any _use case_ for matching 0 ports, it's just that people felt that it was better for the semantics to be _consistent_, even if that meant it was possible to specify something that wasn't very useful.

So anyway, `MinItems=1` is fine. People might come up with reasons why you would want to write "match 0 ports", but we didn't design that in _intentionally_. We just didn't want to special-case the empty list, and didn't consider the fact that we could just forbid it.

**Comment by mikemorris** [comment]
I think this is resolved and can be closed now with @npinaeva's changes in #250?

**Comment by npinaeva** [comment]
@mikemorris good point, added a link from the PR, but I don't have permissions to close the issue

**Comment by bowei** [comment]
Looks like this still might be an issue -- we have to figure out what to do here as it is incompatible with the protobuf generated APIs.

**Comment by bowei** [comment]
Any alternative that is not too hacky:

```
Ports: {
  PortList: []
  # in the future?
  PortRanges: []
  OtherFancyPortStuff
}
```

**Comment by fasaxc** [comment]
@danwinship we discussed this on the call and reopened; the protobuf compiler doesn't like pointers to slices.  Is there anything stopping us from using:

* `nil` slice to mean "do not match on ports"
* `MinItems=1` (as we now added) to prevent the (confusing) "match an empty list of ports"

**Comment by npinaeva** [comment]
May be a good time to re-structure the ports section taking https://github.com/kubernetes-sigs/network-policy-api/issues/187 into account

**Comment by danwinship** [comment]
Do we know that

```go
// +optional
// +kubebuilder:validation:MinItems=1
Ports []AdminNetworkPolicyPort
```

will actually work? ie, it recognizes `nil` and `[]` as distinct values and allows the former (because `optional` but not the latter (becase `MinItems`)?

If it works, then yes, that's OK. Though as Nadia said, we may want to look at larger reorgs anyway

**Comment by bowei** [comment]
I think it's dangerous to assume nil and [] are actually different? It is to a certain extent in Golang but not clear this is actually true in other programming languages. For example, in Golang, all list operations treat nil and [] as equivalent to the extent possible.

**Comment by danwinship** [comment]
> I think it's dangerous to assume nil and [] are actually different?

That's why we made it a pointer to slice.

**Comment by fasaxc** [comment]
> I think it's dangerous to assume nil and [] are actually different?

Right, but now we've forbidden `[]` with MinLength=1, we can treat all zero length slices as meaning "don't match on ports" because the only zero length slice we'll see is the `nil` slice, and we want `nil` to mean "don't match on this".

I don't feel super strongly on it, but adding lots of extra pointer-to-one-element struct makes the API messy and harder to work with.  If it's the k8s norm for this scenario I guess we should follow that.

**Comment by bowei** [comment]
This was posted on Slack thread, but the new validation that is coming in core Kubernetes will disallow pointer-to-slice.

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/code-generator/cmd/validation-gen/validation.go#L315-L324

**Comment by danwinship** [comment]
@fasaxc the question is, will the MinItems=1 make it illegal to write `Ports: nil` in some languages, because that language considers that to be equivalent to `Ports: []`

**Comment by tssurya** [comment]
@bowei will put up an NPEP to hash out what would the future of ports and protocols look like (https://github.com/kubernetes-sigs/network-policy-api/issues/187) both these issues will be tackled as part of the same NPEP. This was discussed on the sig-network meeting on 25th March.

**Comment by danwinship** [comment]
> This was posted on Slack thread, but the new validation that is coming in core Kubernetes will disallow pointer-to-slice.

But the CRD codegen has not been changed yet, right? We should try to make sure that it gets synced up with core codegen so other people don't run into this problem

**Comment by fasaxc** [comment]
We made progress on this one at the in-person meeting at Kubecon.  Outcome was that we agreed:

- To drop the pointer, the type will just be a slice
- Treat empty slice and nil/null/absent equally.  There is no difference in protobuf encodings (which may be supported in future).
- We decided that a pointer to struct containing the slice was not worthwhile.  Being able to say "match this zero-length list of ports" is not useful and i would make the yaml more verbose.

**Comment by tssurya** [comment]
will be fixed via https://github.com/kubernetes-sigs/network-policy-api/issues/187
